docs/middleware.md

```diff
@@ -232,6 +232,252 @@ around explicitly, rather than mutating the middleware instance.
 
+## Pure ASGI Middleware
+
+Due to how ASGI was designed, we are able to build a chain of ASGI applications, on which each application calls the next one.
+Each element of the chain is an [`ASGI`](https://asgi.readthedocs.io/en/latest/) application by itself, which per definition, is also a middleware.
+
+This is also an alternative approach in case the limitations of `BaseHTTPMiddleware` are a problem.
+
+### Guiding principles
+
+The most common way to create an ASGI middleware is with a class.
+
+```python
+class ASGIMiddleware:
+    def __init__(self, app):
+        self.app = app
+
+    async def __call__(self, scope, receive, send):
+        await self.app(scope, receive, send)
+```
+
+The middleware above is the most basic ASGI middleware. It just receives an ASGI application as an argument for its constructor, and implements the `__call__` method, with the parameters that the ASGI protocol defines: `scope`, `receive`, and `send`.
+
+The `scope` parameter is a dictionary that contains the information about the connection.
+You can see more about it [here](https://asgi.readthedocs.io/en/latest/specs/index.html).
+
+Both `receive` and `send` parameters are async callables that work in opposite ways. The `send` callable allows the application to send event messages to the client, and the `receive` allows the application to receive event messages from the client.
+
+As an alternative for the class approach, you can also use a function:
+
+```python
+def asgi_middleware():
+    def asgi_decorator(app):
+        @functools.wraps(app)
+        async def wrapped_app(scope, receive, send):
+            await app(scope, receive, send)
+        return wrapped_app
+    return asgi_decorator
+```
+
+!!! note
+    The function pattern is not commonly spread, but you can check a more advanced implementation of it on
+    [asgi-cors](https://github.com/simonw/asgi-cors/blob/10ef64bfcc6cd8d16f3014077f20a0fb8544ec39/asgi_cors.py).
+
+#### `Scope` types
+
+As we mention, the scope holds the information about the connection. There are three types of `scope`s:
+- `lifespan` is a special type of scope that is used for the lifespan of the ASGI application.
+- `http` is a type of scope that is used for HTTP requests.
+- `websocket` is a type of scope that is used for WebSocket connections.
+
+If you want to create a middleware that only runs on HTTP requests, you'd write something like:
+
+```python
+class ASGIMiddleware:
+    def __init__(self, app):
+        self.app = app
+
+    async def __call__(self, scope, receive, send):
+        if scope["type"] == "http":
+            # Do something here!
+        await self.app(scope, receive, send)
```

Comment:
Sure! I thought it would be easier to explain with the other way, but I guess we should teach what we all do...
```suggestion
    async def __call__(self, scope, receive, send):
        if scope["type"] != "http":
            return await self.app(scope, receive, send)
            
        # Do something here!
        await self.app(scope, receive, send)
```